File: JavaScript/src/Internals/Functions/CsharpWeb/Function_URLENCODE.js
import { Function_1 } from '../Function_1.js';
import { Operand } from '../../../Operand.js';

/**
 * Function_URLENCODE
 */
export class Function_URLENCODE extends Function_1 {
    /**
     * @param {FunctionBase[]} funcs
     */
    constructor(funcs) {
        super(funcs);
    }

    get Name() {
        return "UrlEncode";
    }

    /**
     * @param {AlgorithmEngine} engine
     * @param {Function} tempParameter
     * @returns {Operand}
     */
    evaluate(engine, tempParameter) {
        let args1 = this.getText_1(engine, tempParameter);
        if (args1.IsError) { return args1; }
        let s = args1.TextValue;
        let r = encodeURIComponent(s)
            .replace(/%20/g, '+')
            // 与 C# HttpUtility.UrlEncode 一致：hex 输出小写
            .replace(/%[A-F0-9]{2}/g, (m) => m.toLowerCase());
        return Operand.Create(r);
    }
}

File: java/src/main/java/toolgood/algorithm/internals/functions/csharpweb/Function_URLENCODE.java
package toolgood.algorithm.internals.functions.csharpweb;

import java.net.URLEncoder;
import java.nio.charset.Charset;
import java.util.List;
import java.util.function.BiFunction;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_1;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_URLENCODE extends Function_1 {
    public Function_URLENCODE(FunctionBase func1) {
        super(func1);
    }

    @Override
    public String Name() {
        return "UrlEncode";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        Operand args1 = GetText_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }
        String s = args1.TextValue();
        String r = URLEncoder.encode(s, "UTF-8");
        return Operand.Create(r);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType,
            String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.TEXT);
    }
}
